Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -0,0 +1,131 @@
+
+    /// If `true`, allows the write to opt-out of document level validation.
+    public let bypassDocumentValidation: Bool?
+}
+
+/// A protocol indicating write types that can be batched together using `MongoCollection.bulkWrite`.
+public protocol WriteModel {}
+
+/// A model for an `insertOne` command.
+public struct InsertOneModel: WriteModel {
+    /// The `Document` to insert.
+    public let document: Document
+}
+
+/// A model for a `deleteOne` command.
+public struct DeleteOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+}
+
+/// A model for a `deleteMany` command.
+public struct DeleteManyModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+}
+
+/// A model for a `replaceOne` command.
+public struct ReplaceOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// The `Document` with which to replace the matched document.
+    public let replacement: Document
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// A model for an `updateOne` command.
+public struct UpdateOneModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// A `Document` containing update operators.
+    public let update: Document
+
+    /// A set of filters specifying to which array elements an update should apply.
+    public let arrayFilters: [Document]?
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// A model for an `updateMany` command.
+public struct UpdateManyModel: WriteModel {
+    /// A `Document` representing the match criteria.
+    public let filter: Document
+
+    /// A `Document` containing update operators.
+    public let update: Document
+
+    /// A set of filters specifying to which array elements an update should apply.
+    public let arrayFilters: [Document]?
+
+    /// Specifies a collation to use.
+    public let collation: Document?
+
+    /// When `true`, creates a new document if no document matches the query.
+    public let upsert: Bool?
+}
+
+/// The result of a bulk write operation on a `MongoCollection`.
+public struct BulkWriteResult: Decodable {
+    /// Number of documents inserted.
+    public let insertedCount: Int
+
+    /// Map of the index of the operation to the id of the inserted document.
+    public let insertedIds: [Int: AnyBsonValue]
+
+    /// Number of documents matched for update.
+    public let matchedCount: Int
```

Comment:
Related to the above comment, the update and deleted counts originate from the server and won't be available if the write concern was unacknowledged. I think everything apart from `insertedIds` (which can be generated entirely client-side) should be optional.

Reply:
Ah, I just remembered that you do have `insertedIds` here. If that's going to stay, then I think you should probably provide it for unacknowledged writes as well.